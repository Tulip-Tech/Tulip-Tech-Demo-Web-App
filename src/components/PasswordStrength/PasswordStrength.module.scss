@use 'src/styles/variables';
@use 'src/styles/theme';

@mixin strength($strength, $width, $color) {
  &[data-strength='#{$strength}'] {
    .passwordStrengthFill {
      width: $width;
      background: $color;
    }

    .label {
      color: $color;
    }
  }
}

.passwordStrength {
  @include strength(1, 25%, orangered);
  @include strength(2, 50%, orange);
  @include strength(3, 75%, yellowgreen);
  @include strength(4, 100%, green);

  position: relative;
  display: flex;
  align-items: center;
  height: 16px;
  margin: 8px 0;
  font-size: 14px;
}

.passwordStrengthBar {
  position: relative;
  width: 170px;
  height: 6px;
  margin-right: 8px;

  background: #ddd;
  border-radius: 5px;
}

.passwordStrengthFill {
  position: absolute;
  width: 0;
  height: 100%;
  background: transparent;
  border-radius: inherit;
  transition: width 0.5s ease-in-out, background 0.25s;
}

.label {
  font-weight: theme.$body-font-weight-bold;
  font-size: 14px;
}
