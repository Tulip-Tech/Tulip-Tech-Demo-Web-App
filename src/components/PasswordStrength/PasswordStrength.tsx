import React from 'react';
import { useTranslation } from 'react-i18next';

import styles from './PasswordStrength.module.scss';

type Props = {
  password: string;
};

const PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[0-9]).{8,}$/;

const PasswordStrength: React.FC<Props> = ({ password }: Props) => {
  const { t } = useTranslation('account');

  const passwordStrength = (password: string) => {
    let strength = 0;

    if (!password.match(PASSWORD_REGEX)) return strength;

    if (password.match(/[A-Z]+/)) {
      strength += 1;
    }

    if (password.match(/(\d.*\d)/)) {
      strength += 1;
    }

    if (password.match(/[!,@#$%^&*?_~]/)) {
      strength += 1;
    }

    if (password.match(/([!,@#$%^&*?_~].*[!,@#$%^&*?_~])/)) {
      strength += 1;
    }

    return strength;
  };
  const strength = passwordStrength(password);
  const labels = [
    t('registration.password_strength.invalid'),
    t('registration.password_strength.weak'),
    t('registration.password_strength.fair'),
    t('registration.password_strength.strong'),
    t('registration.password_strength.very_strong'),
  ];

  if (!strength) return null;

  return (
    <div className={styles.passwordStrength} data-strength={strength}>
      <div className={styles.passwordStrengthBar}>
        <div className={styles.passwordStrengthFill} />
      </div>{' '}
      <span className={styles.label}>{labels[strength]}</span>
    </div>
  );
};

export default PasswordStrength;
