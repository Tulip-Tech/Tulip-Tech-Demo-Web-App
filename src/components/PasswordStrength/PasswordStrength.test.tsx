import React from 'react';
import { render } from '@testing-library/react';

import PasswordStrength from './PasswordStrength';

describe('<PasswordStrength>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(<PasswordStrength password="Welcome03" />);

    expect(container).toMatchSnapshot();
  });
});
