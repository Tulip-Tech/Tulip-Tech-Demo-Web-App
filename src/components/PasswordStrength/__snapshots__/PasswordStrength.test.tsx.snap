// Vitest Snapshot v1

exports[`<PasswordStrength> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_passwordStrength_857746"
    data-strength="2"
  >
    <div
      class="_passwordStrengthBar_857746"
    >
      <div
        class="_passwordStrengthFill_857746"
      />
    </div>
     
    <span
      class="_label_857746"
    >
      registration.password_strength.fair
    </span>
  </div>
</div>
`;
